### ISSUE TITLE
modules that can use multiple keyspaces or roles as input

### ISSUE THREAD

**Issue by ivnovi** [ansible-collections/community.cassandra#222]
##### SUMMARY
consider implementing modules that accept multiple items be it keyspaces or roles/users. 

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
cassandra_keyspace, cassandra_role modules

##### ADDITIONAL INFORMATION
when you have more resources (keyspaces, roles) it takes significant amount of time to run playbook where single task is used to create/update all resources, since each task creates its own session. If you could create/update multiple resources with single module it will work much quicker, especially when verifying existing resources. I'm looking to implement this, and thinking either to implement it here as part of collection or continue with in-house modules we use already.

I'm not sure what is best way to structure configuration for such module so it is useful for more users and whether it is aligned with "ansible way". Example:
```yaml
- name: create all ks at once
  cassandra_keyspaces:
    login_host: ..
    ...
    keyspaces:
      - keyspace: ks1
        data_centres:
          dc1: 3
          dc2: 3
      - keyspace: ks2
        data_centres:
          dc1: 3
          dc2: 3
      ...
  cassandra_roles:
    roles:
      - name: rhys
        password: 'secret'
        state: present
        login: yes
        permissions:
          ks1:
            - "ALL PERMISSIONS"
...
      - name: john
        password: 'secret'
        state: present
        login: yes
        permissions:
          ks2:
            - "SELECT"
```

comparing what currently you should do, and it is not quick when you have more items
```yaml
- name: create all ks at once
  cassandra_keyspace:
    name: "{{ item['name'] }}"
    ...
  with_items: "{{ keyspaces }}"
- name: create all ks at once
  cassandra_role:
    name: "{{ item['name'] }}"
    ...
  with_items: "{{ roles }}"
```

it could be also single module for both so that you could do more checks to verify if keyspace exists before performing grants and similar logic, similar as single module already is used for roles+grants.

**Comment by rhysmeister** [comment]
Hello @ivnovi,

The "Ansible Way" would be to do one operation per module call. I'm not convinved this type of functionality belongs in a module. If you're looking to speed things up you could...

- Use the template module to write a cql file, with all your commands, and then use the cassandra_cqlsh module to execute it.
- Have a look at [mitogen](https://dev.to/sshnaidm/speed-up-ansible-with-mitogen-2c3j) for Ansible. 

I'm not sure how much of an improvement you'll get as I would think a significant amount of the time is spent on communication between the nodes in the C* Cluster. Neither the suggestions above or your idea would do anything in terms of that.

Cheers,

Rhys

**Comment by rhysmeister** [comment]
Use tags on your tasks. If nothing has changed, which you shoudl know, then you can just skip the task completely.

**Comment by ivnovi** [comment]
It does work for us like this right now, but still wastes a lot of time, this is why I'm looking to improve this.  And still as test have shown even if I run only against keyspaces it can be 10x improvement. And if you have multiple clusters to apply it to, it is still adds up.

There are multiple ways to improve this, for example splitting configuration on some criteria and then running the part which changed (subset of keyspaces, roles, etc). This probably can even be done in pipeline based on file that changed, but again, if you can run against all with some tweaks, why not.

**Comment by rhysmeister** [comment]
I think most of the time spent will be from the ssh connection side not directly from C* sessions. You might see a good improvement with [Ansible Pipelining](https://technotes.adelerhof.eu/ansible/pipelining/). Using with_items will probably add a 1-2 second delay between each invocation.

**Comment by ivnovi** [comment]
Actually I realized that I dont currently use ssh connections in my tests, but connect to C* cql port directly. This way I dont care about python on a node, which sometimes is issue. Completing single task takes around 1s, from which 0.3-0.4 is establishing session and getting answer from C*, most of which is creating session, which I got by adding additional logging inside module. 

```
 [
     "start",
     "15:57:05.880977"
 ],
 [
     "session created",
     "15:57:06.245323"
 ],
 [
     "if keyspace exist start",
     "15:57:06.245371"
 ],
 [
     "end of module before exit",
     "15:57:06.253314"
 ],
 [
     "diff",
     "372356"
 ]
```

Pipelining does not make difference, probably because there is no ssh.

With mitogen playbook run produces some additional messages about Protocol error, but correctly identifies necessary change and completes in 25 seconds (comparing to 37s by default), not sure if it is real improvement, or because it kills process at some late stage or similar. If legit it is significant improvement of course, but would need to investigate more about the tool to understand whether it can be used universally, and why it produces those messages.

When I switched my experimental verify module playbook not to use ssh (delegate_to: localhost) it runs even faster, less than 3s
